src/hotspot/share/opto/vectornode.cpp

```diff
@@ -1541,6 +1588,76 @@ Node* VectorLongToMaskNode::Ideal(PhaseGVN* phase, bool can_reshape) {
   return NULL;
 }
 
+// Generate other vector nodes to implement the masked/non-masked vector negation.
+Node* VectorNode::degenerate_vector_negate(Node* n, int vlen, BasicType bt, PhaseGVN* phase, bool is_predicated) {
+  const TypeVect* vt = TypeVect::make(bt, vlen);
+  // Transformation for predicated NegVI/L
+  if (is_predicated) {
+      // (NegVI/L src m) ==> (AddVI/L (XorV src (ReplicateI/L -1) m) (ReplicateI/L 1) m)
+      Node* const_minus_one = NULL;
+      Node* const_one = NULL;
+      int add_opc;
+      if (bt == T_LONG) {
+        const_minus_one = phase->longcon(-1L);
+        const_one = phase->longcon(1L);
+        add_opc = Op_AddL;
+      } else {
+        const_minus_one = phase->intcon(-1);
+        const_one = phase->intcon(1);
+        add_opc = Op_AddI;
+      }
+      const_minus_one = phase->transform(VectorNode::scalar2vector(const_minus_one, vlen, Type::get_const_basic_type(bt)));
+      Node* xorv = VectorNode::make(Op_XorV, n->in(1), const_minus_one, vt);
+      xorv->add_req(n->in(2));
+      xorv->add_flag(Node::Flag_is_predicated_vector);
+      phase->transform(xorv);
+      const_one = phase->transform(VectorNode::scalar2vector(const_one, vlen, Type::get_const_basic_type(bt)));
+      Node* addv = VectorNode::make(VectorNode::opcode(add_opc, bt), xorv, const_one, vt);
+      addv->add_req(n->in(2));
+      addv->add_flag(Node::Flag_is_predicated_vector);
+      return addv;
+  }
+
+  // NegVI/L ==> (SubVI/L (ReplicateI/L 0) src)
+  Node* const_zero = NULL;
+  int sub_opc;
+  if (bt == T_LONG) {
+    const_zero = phase->longcon(0L);
+    sub_opc = Op_SubL;
+  } else {
+    const_zero = phase->intcon(0);
+    sub_opc = Op_SubI;
+  }
+  const_zero = phase->transform(VectorNode::scalar2vector(const_zero, vlen, Type::get_const_basic_type(bt)));
+  return VectorNode::make(VectorNode::opcode(sub_opc, bt), const_zero, n->in(1), vt);
+}
+
+Node* NegVINode::Ideal(PhaseGVN* phase, bool can_reshape) {
```

Comment:
Much duplication in `NegVINode::Ideal` and `NegVLNode::Ideal`.
Is it possible to refactor the implementation?

Reply:
Yeah, maybe we need a superclass for `NegVINode` and `NegVLNode`. Thanks!